File: CSharp/07-explain-and-debug/LegacyParser.cs
// Demo 7 — Explain Code & Debug
// ===============================
// This is "legacy" code that works but is hard to understand.
// Use Copilot to explain it and then improve it.
//
// Exercises:
//   1. Select ParseLogEntry() → Chat → "Explain this method step by step"
//   2. Select the regex → Inline Chat → "What does this regex match?"
//   3. Ask Chat: "Rewrite this using named groups and make it readable"
//   4. Select AnalyzeLogFile() → Chat → "This is slow for large files. Find the issue."
//   5. Run the file and inspect the output.

using System.Text.RegularExpressions;

namespace ExplainAndDebug;

public class LogAnalyzer
{
    private static readonly Regex LogPattern = new(
        @"(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2})"   // timestamp
      + @"\s+\[(\w+)\]"                                // log level
      + @"\s+(\S+)"                                     // source module
      + @"\s+-\s+(.*)",                                 // message
        RegexOptions.Compiled
    );

    public LogEntry? ParseLogEntry(string line)
    {
        // Ask Copilot: "Explain what this method does and how the regex works."
        var match = LogPattern.Match(line.Trim());
        if (!match.Success) return null;

        return new LogEntry
        {
            Timestamp = DateTime.Parse(match.Groups[1].Value),
            Level = match.Groups[2].Value,
            Source = match.Groups[3].Value,
            Message = match.Groups[4].Value
        };
    }

    public LogReport AnalyzeLogFile(IEnumerable<string> lines)
    {
        // Ask Copilot: "This is O(n²) in a subtle way — can you find and fix it?"
        var entries = new List<LogEntry>();
        foreach (var line in lines)
        {
            var parsed = ParseLogEntry(line);
            if (parsed is not null)
                entries.Add(parsed);
        }

        // Count by level
        var levelCounts = entries
            .GroupBy(e => e.Level)
            .ToDictionary(g => g.Key, g => g.Count());

        // Count by source
        var sourceCounts = entries
            .GroupBy(e => e.Source)
            .ToDictionary(g => g.Key, g => g.Count());

        // Find error messages
        var errors = entries
            .Where(e => e.Level == "ERROR")
            .Select(e => e.Message)
            .ToList();

        // Time range
        var timestamps = entries.Select(e => e.Timestamp).ToList();
        var timeRange = timestamps.Count > 0
            ? (Start: timestamps.Min(), End: timestamps.Max())
            : ((DateTime Start, DateTime End)?)null;

        // ⚠️ Subtle issue: duplicate detection is O(n²)
        // Ask Copilot to spot and fix it.
        var duplicates = new List<string>();
        for (int i = 0; i < entries.Count; i++)
        {
            for (int j = i + 1; j < entries.Count; j++)
            {
                if (entries[j].Message == entries[i].Message)
                {
                    if (!duplicates.Contains(entries[i].Message))
                        duplicates.Add(entries[i].Message);
                }
            }
        }

        return new LogReport
        {
            TotalEntries = entries.Count,
            LevelCounts = levelCounts,
            SourceCounts = sourceCounts,
            Errors = errors,
            TimeRange = timeRange,
            DuplicateMessages = duplicates
        };
    }
}

public class LogEntry
{
    public DateTime Timestamp { get; init; }
    public string Level { get; init; } = "";
    public string Source { get; init; } = "";
    public string Message { get; init; } = "";
}

public class LogReport
{
    public int TotalEntries { get; init; }
    public Dictionary<string, int> LevelCounts { get; init; } = new();
    public Dictionary<string, int> SourceCounts { get; init; } = new();
    public List<string> Errors { get; init; } = new();
    public (DateTime Start, DateTime End)? TimeRange { get; init; }
    public List<string> DuplicateMessages { get; init; } = new();
}

// ─────────────────────────────────────────────────────────────────────
// Sample log data for testing
// ─────────────────────────────────────────────────────────────────────
public class Program
{
    private static readonly string[] SampleLogs =
    [
        "2026-06-10T09:00:01 [INFO] auth.service - User alice@example.com logged in",
        "2026-06-10T09:00:03 [DEBUG] db.pool - Connection acquired from pool (active: 5)",
        "2026-06-10T09:00:05 [INFO] api.orders - GET /api/orders — 200 OK (23ms)",
        "2026-06-10T09:00:07 [WARN] api.orders - Slow query detected: 1200ms on orders table",
        "2026-06-10T09:00:09 [ERROR] payment.gateway - Timeout connecting to Stripe API",
        "2026-06-10T09:00:12 [INFO] auth.service - User bob@example.com logged in",
        "2026-06-10T09:00:15 [ERROR] payment.gateway - Timeout connecting to Stripe API",
        "2026-06-10T09:00:18 [INFO] api.orders - POST /api/orders — 201 Created (45ms)",
        "2026-06-10T09:00:20 [DEBUG] db.pool - Connection released back to pool (active: 4)",
        "2026-06-10T09:00:22 [ERROR] api.orders - Unhandled exception in order validation",
    ];

    public static void Main()
    {
        var analyzer = new LogAnalyzer();
        var report = analyzer.AnalyzeLogFile(SampleLogs);

        Console.WriteLine($"Total entries: {report.TotalEntries}");
        Console.WriteLine($"By level: {string.Join(", ", report.LevelCounts.Select(kv => $"{kv.Key}={kv.Value}"))}");
        Console.WriteLine($"By source: {string.Join(", ", report.SourceCounts.Select(kv => $"{kv.Key}={kv.Value}"))}");
        Console.WriteLine($"Errors: {string.Join("; ", report.Errors)}");
        if (report.TimeRange.HasValue)
            Console.WriteLine($"Time range: {report.TimeRange.Value.Start:s} → {report.TimeRange.Value.End:s}");
        Console.WriteLine($"Duplicates: {string.Join("; ", report.DuplicateMessages)}");
    }
}

File: Java/07-explain-and-debug/LegacyParser.java
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * Demo 7 — Explain Code & Debug (Refactored to Java)
 * ====================================================
 * This is "legacy" code that works but is hard to understand.
 * Use Copilot to explain it and then improve it.
 *
 * Exercises:
 *   1. Select parseLogEntry() → Chat → "Explain this function step by step"
 *   2. Select the regex pattern → Inline Chat → "What does this regex match?"
 *   3. Ask Chat: "Rewrite this using named groups and make it readable"
 *   4. Select analyzeLogFile() → Chat → "This is slow for large files. Optimize it."
 *   5. Run the file and inspect the output.
 */

/**
 * Represents a single log entry with structured data.
 */
class LogEntry {
    private LocalDateTime timestamp;
    private String level;
    private String source;
    private String message;

    public LogEntry(LocalDateTime timestamp, String level, String source, String message) {
        this.timestamp = timestamp;
        this.level = level;
        this.source = source;
        this.message = message;
    }

    public LocalDateTime getTimestamp() {
        return timestamp;
    }

    public String getLevel() {
        return level;
    }

    public String getSource() {
        return source;
    }

    public String getMessage() {
        return message;
    }

    @Override
    public String toString() {
        return "LogEntry{" +
                "timestamp=" + timestamp +
                ", level='" + level + '\'' +
                ", source='" + source + '\'' +
                ", message='" + message + '\'' +
                '}';
    }
}

/**
 * Log analysis result containing statistics about the log entries.
 */
class LogAnalysisReport {
    private int totalEntries;
    private Map<String, Integer> levelCounts;
    private Map<String, Integer> sourceCounts;
    private List<String> errors;
    private TimeRange timeRange;
    private List<String> duplicateMessages;

    public LogAnalysisReport(
        int totalEntries,
        Map<String, Integer> levelCounts,
        Map<String, Integer> sourceCounts,
        List<String> errors,
        TimeRange timeRange,
        List<String> duplicateMessages) {
        this.totalEntries = totalEntries;
        this.levelCounts = levelCounts;
        this.sourceCounts = sourceCounts;
        this.errors = errors;
        this.timeRange = timeRange;
        this.duplicateMessages = duplicateMessages;
    }

    public int getTotalEntries() {
        return totalEntries;
    }

    public Map<String, Integer> getLevelCounts() {
        return levelCounts;
    }

    public Map<String, Integer> getSourceCounts() {
        return sourceCounts;
    }

    public List<String> getErrors() {
        return errors;
    }

    public TimeRange getTimeRange() {
        return timeRange;
    }

    public List<String> getDuplicateMessages() {
        return duplicateMessages;
    }

    @Override
    public String toString() {
        return "LogAnalysisReport{" +
                "totalEntries=" + totalEntries +
                ", levelCounts=" + levelCounts +
                ", sourceCounts=" + sourceCounts +
                ", errors=" + errors +
                ", timeRange=" + timeRange +
                ", duplicateMessages=" + duplicateMessages +
                '}';
    }
}

/**
 * Represents a time range with start and end timestamps.
 */
class TimeRange {
    private String start;
    private String end;

    public TimeRange(String start, String end) {
        this.start = start;
        this.end = end;
    }

    public String getStart() {
        return start;
    }

    public String getEnd() {
        return end;
    }

    @Override
    public String toString() {
        return "TimeRange{" +
                "start='" + start + '\'' +
                ", end='" + end + '\'' +
                '}';
    }
}

public class LegacyParser {

    private static final Pattern LOG_PATTERN = Pattern.compile(
        "(\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2})" +  // timestamp
        "\\s+\\[(\\w+)\\]" +                             // log level
        "\\s+(\\S+)" +                                    // source module
        "\\s+-\\s+(.*)"                                   // message
    );

    private static final DateTimeFormatter TIMESTAMP_FORMATTER = DateTimeFormatter.ISO_LOCAL_DATE_TIME;

    /**
     * Parse a single log line into structured data.
     * Ask Copilot: 'Explain what this function does and how the regex works.'
     *
     * @param line the log line to parse
     * @return a LogEntry if parsing succeeds, null otherwise
     */
    public static LogEntry parseLogEntry(String line) {
        Matcher m = LOG_PATTERN.matcher(line.strip());
        if (!m.matches()) {
            return null;
        }

        String timestamp = m.group(1);
        String level = m.group(2);
        String source = m.group(3);
        String message = m.group(4);

        LocalDateTime ts = LocalDateTime.parse(timestamp, TIMESTAMP_FORMATTER);
        return new LogEntry(ts, level, source, message);
    }

    /**
     * Analyze a list of log lines and produce statistics.
     * Ask Copilot: 'This has an O(n²) duplicate detection — can you find and fix it?'
     *
     * @param lines the list of log lines to analyze
     * @return a LogAnalysisReport with statistics
     */
    public static LogAnalysisReport analyzeLogFile(List<String> lines) {
        // Parse all log entries
        List<LogEntry> entries = lines.stream()
            .map(LegacyParser::parseLogEntry)
            .filter(Objects::nonNull)
            .collect(Collectors.toList());

        // Count by level
        Map<String, Integer> levelCounts = new HashMap<>();
        for (LogEntry e : entries) {
            levelCounts.merge(e.getLevel(), 1, Integer::sum);
        }

        // Count by source
        Map<String, Integer> sourceCounts = new HashMap<>();
        for (LogEntry e : entries) {
            sourceCounts.merge(e.getSource(), 1, Integer::sum);
        }

        // Find error messages
        List<String> errors = entries.stream()
            .filter(e -> "ERROR".equals(e.getLevel()))
            .map(LogEntry::getMessage)
            .collect(Collectors.toList());

        // Time range
        TimeRange timeRange = null;
        if (!entries.isEmpty()) {
            LocalDateTime minTime = entries.stream()
                .map(LogEntry::getTimestamp)
                .min(LocalDateTime::compareTo)
                .orElse(null);
            LocalDateTime maxTime = entries.stream()
                .map(LogEntry::getTimestamp)
                .max(LocalDateTime::compareTo)
                .orElse(null);
            if (minTime != null && maxTime != null) {
                timeRange = new TimeRange(minTime.toString(), maxTime.toString());
            }
        }

        // ⚠️ Subtle issue: duplicate detection is O(n²)
        // Ask Copilot to spot and fix it.
        Set<String> duplicateSet = new HashSet<>();
        Set<String> seenMessages = new HashSet<>();
        for (LogEntry entry : entries) {
            String msg = entry.getMessage();
            if (seenMessages.contains(msg)) {
                duplicateSet.add(msg);
            }
            seenMessages.add(msg);
        }
        List<String> duplicates = new ArrayList<>(duplicateSet);

        return new LogAnalysisReport(
            entries.size(),
            levelCounts,
            sourceCounts,
            errors,
            timeRange,
            duplicates
        );
    }

    // ─────────────────────────────────────────────────────────────────────
    // Sample log data for testing
    // ─────────────────────────────────────────────────────────────────────

    private static final List<String> SAMPLE_LOGS = Arrays.asList(
        "2026-06-08T09:00:01 [INFO] auth.service - User alice@example.com logged in",
        "2026-06-08T09:00:03 [DEBUG] db.pool - Connection acquired from pool (active: 5)",
        "2026-06-08T09:00:05 [INFO] api.orders - GET /api/orders — 200 OK (23ms)",
        "2026-06-08T09:00:07 [WARN] api.orders - Slow query detected: 1200ms on orders table",
        "2026-06-08T09:00:09 [ERROR] payment.gateway - Timeout connecting to stripe API",
        "2026-06-08T09:00:12 [INFO] auth.service - User bob@example.com logged in",
        "2026-06-08T09:00:15 [ERROR] payment.gateway - Timeout connecting to stripe API",
        "2026-06-08T09:00:18 [INFO] api.orders - POST /api/orders — 201 Created (45ms)",
        "2026-06-08T09:00:20 [DEBUG] db.pool - Connection released back to pool (active: 4)",
        "2026-06-08T09:00:22 [ERROR] api.orders - Unhandled exception in order validation"
    );

    public static void main(String[] args) {
        LogAnalysisReport report = analyzeLogFile(SAMPLE_LOGS);

        System.out.println("Total entries: " + report.getTotalEntries());
        System.out.println("By level: " + report.getLevelCounts());
        System.out.println("By source: " + report.getSourceCounts());
        System.out.println("Errors: " + report.getErrors());
        System.out.println("Time range: " + report.getTimeRange());
        System.out.println("Duplicates: " + report.getDuplicateMessages());
    }
}
